playback/src/player.rs

```diff
@@ -145,33 +165,32 @@ impl PlayerState {
     fn decoder(&mut self) -> Option<&mut Decoder> {
         use self::PlayerState::*;
         match *self {
-            Stopped => None,
+            Stopped | EndOfTrack { .. } => None,
             Paused { ref mut decoder, .. } |
             Playing { ref mut decoder, .. } => Some(decoder),
             Invalid => panic!("invalid state"),
         }
     }
 
-    fn signal_end_of_track(self) {
+    fn playing_to_end_of_track(&mut self) {
         use self::PlayerState::*;
-        match self {
-            Paused { end_of_track, .. } |
-            Playing { end_of_track, .. } => {
-                let _ = end_of_track.send(());
-            }
-
-            Stopped => warn!("signal_end_of_track from stopped state"),
-            Invalid => panic!("invalid state"),
+        match mem::replace(self, Invalid) {
+            Playing { track_id, end_of_track, ..} => {
+                end_of_track.send(());
+                let old_state = mem::replace(self, EndOfTrack { track_id });
```

Comment:
This can just be `*self = EndOfTrack { ... }`

Reply:
Fixed!